AutomatedTesting/cmake/EngineFinder.cmake

```diff
@@ -36,6 +36,32 @@ if(CMAKE_MODULE_PATH)
     endforeach()
 endif()
 
+# Read the user/project_settings.json file and look for the engine_path
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/user/project_settings.json)
+    file(READ ${CMAKE_CURRENT_SOURCE_DIR}/user/project_settings.json project_settings_json)
+    string(JSON engine_path ERROR_VARIABLE json_error GET ${project_settings_json} "engine_path")
+    if(json_error)
+        string(CONCAT project_registration_error "Unable to read key 'engine_path' from "
+            "'${CMAKE_CURRENT_SOURCE_DIR}/user/project_settings.json'\n"
+            "Error: ${json_error}\n"
+            "Project registration is required before configuring. "
+            "Register this project by adding it to an engine in Project Manager or run "
+            "the following command from the engine root:\n"
+            "scripts/o3de register --pp \"${CMAKE_CURRENT_SOURCE_DIR}\"\n")
+        message(FATAL_ERROR "${project_registration_error}")
+    elseif(EXISTS ${engine_path}/cmake)
+        message(INFO " Using engine from user/project_settings.json ${engine_path}")
+        list(APPEND CMAKE_MODULE_PATH "${engine_path}/cmake")
+        return()
+    else()
+        string(CONCAT engine_path_info "Unable to use the engine_path from "
+            "'${CMAKE_CURRENT_SOURCE_DIR}/user/project_settings.json' because "
+            "'${engine_path}/cmake' does not exist.\n"
+            "Falling back to look for an engine named '${LY_ENGINE_NAME_TO_USE}' in o3de_manifest.json")
+        message(INFO " ${engine_path_info}")
+    endif()
+endif()
```

Comment:
I think we should reverse how CMake finds the engine after a project upgrade or for any new projects.

Instead of using the EngineFinder.cmake to initially to locate the engine, I think the project.json should have a "engine_finder" key that points to a path to an "engine finder cmake" script. Any relative paths would be treated relative to the `<project-root>` and absolute paths would be used as is.
Side Note: The [cmake_path](https://cmake.org/cmake/help/latest/command/cmake_path.html?highlight=cmake_path#normalization) takes care of this automatically as it works with similarly to python [os.path.join](https://docs.python.org/3/library/os.path.html#os.path.join) where an absolute path segment discards any previous path segments so far.

### project.json
```
{
    "engine_finder": "cmake/EngineFinder.cmake"
}
```

Now for the naming of the `project_settings.json` file. I think it should just be called `project.json` and it is used as an overlay for the `<project-root>/project.json` by merging its settings with it.
i.e It would work similar to the Settings Registry with how the `<project-root>/user/Registry` overlays the `<project-root>/Registry` directory.

Pretty much it can look for two files currently
```
<project-root>/project.json
<project-root>/user/project.json
```

Instead of allowing the user to specify an engine path, they can specify an "override" for the "engine" key identifier to locate an alternative registered engine.

Reply:
By using an overlay of settings for the project.json this allows team and individual contributors to specify there engine to use separately from each other.

Now back to the "engine_finder" cmake key, this can be used provide upgraded Find engine scripts for the project, without needing to overwrite the existing `cmake/EngineFinder.cmake` in the project. This is useful in case the upgrade as failed and the maintainer checks in the new files `cmake/EngineFinder.cmake` files to source control, only for user to discover an issue with it.
Instead the upgrade back could be to add a `cmake/EngineFinder_v2.cmake` script to their project and update their project.json to start using that particular file.

Reply:
I like it, so I'm assuming we'd update the project `CMakeList.txt` to do the initial loading of `project.json` and overlaying with `user/project.json` and then including the specified `engine_finder_cmake` entry?

Reply:
Yes the project CMakeLists.txt would read the `project.json` and then the `user/project.json` for the "engine_finder_cmake" entry and then `include()` that path.